src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataDictionaryFactory.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Linq.Expressions;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    public static class ViewDataDictionaryFactory
+    {
+        internal delegate object ViewDataDictionaryCreator(object[] arguments);
+
+        private static readonly ConcurrentDictionary<ConstructorInfo, ViewDataDictionaryCreator> 
+            _viewDataDictionaryCreateDelegateCache = new ConcurrentDictionary<ConstructorInfo, ViewDataDictionaryCreator>();
```

Comment:
We don't want to add another (separate) cache if possible. RazorPageActivator already has a cache - just include this with the existing data.